### PR TITLE
feat: :construction_worker: add a redux store showcase

### PR DESCRIPTION
**Pull request by AdrianValdes**
I created something a little different than what you asked because I wanted to test new features in `redux`, like `useDispatch()`, and `useSelector()`. Basically, I wanted to practice how to use `redux` with `hooks`.  

I also added more boilerplate than what we need for an example like this one. I did it in order to set a good basis and to have a better idea of how or `redux store` would like and how to use it with Typescript.

To do:
  1. Decide a folder structure for redux and for the app in general (I moved App.ts inside the `components` folder, but just to not have it in the root).

  2. Decide if we are using redux-toolkit or not. (I know you told me to use it if I wanted, but I am not sure if that is the best to do in our case, can we talk about it?)

### CODE REVIEW

**Comment by AdrianValdes** [comment]
It is very weird that the Github Actions test are failing now. I checked and it is because the `bootstrap` action is trying to find `eslint-config-small-ads` in npm

>  https://registry.npmjs.org/eslint-config-small-ads

And it says that it cannot find it. Did we change something? Because that was working fine. I also tried  locally and all works fine.

**Comment by paschalidi** [comment]
I get the same error locally. Can you look into it?

Usually you wont be expecting other developers to start looking in a PR that is failing. We have the checks here so that we make sure everything works before we start reviewing. Of course this is not the case when you really stuck and you need help from another developer solving an issue on the gh checks.

**Comment by paschalidi** [comment]
I believe this is a problem with the `package.lock.json`. What you can do is delete all the `package.lock.json` remove all the node modules from each project. 

Then in the root you will be doing a "clean" install, with `npm install` and then `npm run bootstrap`. 
This will first install the dependencies for the root, and the second command will install/link dependencies for the modules we have.

**Comment by paschalidi** [comment]
Also about the points you making. For the first 
>Decide a folder structure for redux and for the app in general (I moved App.ts inside the components folder, but just to not have it in the root).

Lets discuss it yes.

For the 2.
> Decide if we are using redux-toolkit or not. (I know you told me to use it if I wanted, but I am not sure if that is the best to do in our case, can we talk about it?)

Mind explaining the what disadvantages you see in using `redux-toolkit`. This is open for discussion, lets move the discussion by hearing your concerns

File: src/fe-small-advertisements/package.json [hunk 1]
```diff
@@ -10,9 +10,13 @@
     "@types/node": "^12.19.14",
     "@types/react": "^16.14.2",
     "@types/react-dom": "^16.9.10",
+    "@types/react-redux": "^7.1.16",
     "react": "^17.0.1",
     "react-dom": "^17.0.1",
+    "react-redux": "^7.2.2",
     "react-scripts": "4.0.1",
+    "redux": "^4.0.5",
+    "redux-devtools-extension": "^2.13.8",
```
Comment (paschalidi) [inline src/fe-small-advertisements/package.json:19]: We want this to be a `devDependency`. When we build the project we dont want to have extra packages. This package we use only for development. Can you look into it a little please?

File: src/fe-small-advertisements/src/App.tsx [hunk 2]
```diff
@@ -1,11 +0,0 @@
...
-
-function App() {
-  return (
-    <>
-      <h1>Hello World</h1>
-    </>
-  );
-}
-
-export default App;
```
Comment (paschalidi) [inline src/fe-small-advertisements/src/App.tsx:11]: Lets leave the app where it used to be. Later in this file we will be doing our routing. Or if you insist moving it. Make a folder for it and also move the `App.test.tsx` in this folder. Now the app and the tests of it are in different places.

File: src/fe-small-advertisements/src/actions/index.ts [hunk 3]
```diff
@@ -0,0 +1,2 @@
+export * from './name';
```
Comment (paschalidi) [inline src/fe-small-advertisements/src/actions/index.ts:1]: `name` is such a generic name. Can we use something more specific in this case?

File: src/fe-small-advertisements/src/actions/types.ts [hunk 4]
```diff
@@ -0,0 +1,3 @@
+import { ChangeNameAction } from './name';
+
+export type UserAction = ChangeNameAction;
```
Comment (paschalidi) [inline src/fe-small-advertisements/src/actions/types.ts:3]: I dont het this? Why weneed to have a `ChangeNameAction` _and_ a `UserAction` if they both the same?

File: src/fe-small-advertisements/src/actions/name.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+export interface ChangeNameAction {
+  type: typeof CHANGE_NAME;
+  payload: User;
+}
+
+export const changeNameActionCreator = ({
+  name,
+}: {
+  name: string;
+}): ChangeNameAction => ({ type: CHANGE_NAME, payload: { name } });
```
Comment (paschalidi) [inline src/fe-small-advertisements/src/actions/name.ts:16]:
```suggestion
}: User): ChangeNameAction => ({ type: CHANGE_NAME, payload: { name } });
```
isnt this `User`?

File: src/fe-small-advertisements/src/components/App.tsx [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+import { useDispatch, useSelector } from 'react-redux';
+import { changeNameActionCreator } from '../actions';
+import { RootState } from '../reducers';
+
+function App() {
+  const { name } = useSelector((state: RootState) => state.user);
+  const dispatch = useDispatch();
+  const [newName, setNewName] = useState('');
+
+  const handleNewName = (e: ChangeEvent<HTMLInputElement>): void => {
```
Comment (paschalidi) [inline src/fe-small-advertisements/src/components/App.tsx:11]:
```suggestion
  const setNewName = (e: ChangeEvent<HTMLInputElement>): void => {
```

You basically setting a new name.

File: src/fe-small-advertisements/src/actions/name.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+}
+
+export const CHANGE_NAME = 'CHANGE_NAME';
+
+export interface ChangeNameAction {
+  type: typeof CHANGE_NAME;
+  payload: User;
+}
+
+export const changeNameActionCreator = ({
```
Comment (paschalidi) [inline src/fe-small-advertisements/src/actions/name.ts:12]: Avoid using `ActionCreator` as suffix. Its gonna be used as argument in a dispatch function like this: 
```
dispatch(changeNameActionCreator({ name: newName }));
```

We know it is an action creator by the fact that is consumed from the `dispatch`

File: src/fe-small-advertisements/src/components/App.tsx [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    setNewName(e.target.value);
+  };
+
+  const handleUpdateTitle = (e: FormEvent<HTMLFormElement>): void => {
+    e.preventDefault();
+    dispatch(changeNameActionCreator({ name: newName }));
```
Comment (paschalidi) [inline src/fe-small-advertisements/src/components/App.tsx:17]: Good thinking using object with key `name` here.

File: src/fe-small-advertisements/src/components/App.tsx [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+      <p>My name is: {name}</p>
+      <form onSubmit={handleUpdateTitle}>
+        <input value={newName} onChange={handleNewName} />
+        <button type="submit">change name</button>
+      </form>
+    </>
+  );
+}
+
+export default App;
```
Comment (paschalidi) [inline src/fe-small-advertisements/src/components/App.tsx:32]: I think we agreed these will be `const`, mind swapping it?

File: src/fe-small-advertisements/src/index.tsx [hunk 10]
```diff
@@ -1,12 +1,20 @@
...
+import { Provider } from 'react-redux';
+import { createStore } from 'redux';
+import { composeWithDevTools } from 'redux-devtools-extension';
 
-import App from './App';
+import App from './components/App';
+import { rootReducer } from './reducers';
 import reportWebVitals from './reportWebVitals';
 
+const store = createStore(rootReducer, composeWithDevTools());
```
Comment (paschalidi) [inline src/fe-small-advertisements/src/index.tsx:11]: You need ti find a way to add this middleware to the store only when we are on `localhost`

**Review by paschalidi — CHANGES_REQUESTED** [review]
Looks good! Have some comments that you might wanna be looking at. 

Also, you asked me about the structure. Can you look up and read about domain driven structure on project. I would suggest we go that direction for the `webapp`

About the `redux-toolkit` this is still pending and is best if we make a decision in this PR.

Please fix all the comments you think are important and I will have another look then )

File: src/fe-small-advertisements/src/components/App.tsx [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+import { RootState } from '../reducers';
+
+function App() {
+  const { name } = useSelector((state: RootState) => state.user);
+  const dispatch = useDispatch();
+  const [newName, setNewName] = useState('');
+
+  const handleNewName = (e: ChangeEvent<HTMLInputElement>): void => {
+    setNewName(e.target.value);
+  };
```
Comment (paschalidi) [inline src/fe-small-advertisements/src/components/App.tsx:13]: I dont think we need this function here. Can you try use `setNewName` directly on the `onChange`?

File: src/fe-small-advertisements/src/actions/name.ts (hunk 7 above)
Comment (paschalidi) [inline src/fe-small-advertisements/src/actions/name.ts:12]: Also I think here the name could be improved a little. You are changing the name indeed but actually you are setting the name. Something like `setName` might be more descriptive? 
This is only a suggestion, `changeName` is perhaps alright, so feel free to do as you please

**Comment by AdrianValdes** [comment]
> I get the same error locally. Can you look into it?
> 
> Usually you wont be expecting other developers to start looking in a PR that is failing. We have the checks here so that we make sure everything works before we start reviewing. Of course this is not the case when you really stuck and you need help from another developer solving an issue on the gh checks.

You are totally right, I understand it should be my job to fix those issues. I was super confused because the tests were not failing locally for me. I tried to solve it and I couldn't. But I think I didn't try enough, my bad.

**Comment by AdrianValdes** [comment]
> Mind explaining the what disadvantages you see in using redux-toolkit. This is open for discussion, lets move the discussion by hearing your concerns

My concerns were basically two:

1. Using redux-toolkit were abstract the `redux` logic so much that could be hard for me to follow what is doing.
2. Using the `immer` to "fake-mutate" the data directly kind of weird to me now. I am so used to avoid modifying the state that it feels "funny" to not do it manually anymore.

Nevertheless, after you asked me to explain the disadvantages, I convinced myself that it is indeed "better" to use `redux-toolkit`. The reasons:

1. It is a lot less code, A LOT.
2. It is not that hard to follow what is doing.
3. Use the functionality the `immer` provides is optional. I could always keep doing what I was doing. (Nevertheless, I realize we have to decide for one way of dealing with the data in the reducers, I will use what you think is best)

> About the `redux-toolkit` this is still pending and is best if we make a decision in this PR.

Because you were in favor of using `redux-toolkit` from the beginning and now I am convinced too, I will approach all your comments and will refactor the code to use ´redux-toolkit`. It is right if I do so?

**Comment by paschalidi** [comment]
Yep you can do it that way 👍. If in the future we have different opinion it is good practise to try to raise your opinion and gather arguments why you think something might be better another way. I dont wanna discourage you from raising your opinion, even though I might seem like I have a strong opinion on something I will be always listening to your points when they come structured and logical :) So keep up the good work and is always good to have your own opinion about how things would work

File: src/fe-small-advertisements/package.json (hunk 1 above)
Comment (AdrianValdes) [reply src/fe-small-advertisements/package.json:19]: Yes, you are right. It makes sense, I actually thought about this when I added it. But then I read that some people actually leave it in production. Anyways, it make more sense what you are saying.

File: src/fe-small-advertisements/src/App.tsx (hunk 2 above)
Comment (AdrianValdes) [reply src/fe-small-advertisements/src/App.tsx:11]: I moved it back. I really have to read about folder structure, I already started to read about DDD and I was reading about `features-folder` because that is the way the official docs of `redux-toolkit` recommend.

> Note that while it doesn't matter how you actually structure your folders and files, we've found that a "feature folder" approach usually works better for most applications. The file name is also entirely up to you, but a convention of someFeatureSlice.js is reasonable to use.

For this commit  I created a /features/person folder and put the personSlice.ts there. I just to put it somewhere until I read more about DDD, is that fine? Or is that sloppy of me?

 Sometimes I am not sure what to do in cases like this one: On the one hand, you asked me to build just a showcase and I have to decide how to make that showcase the best way, but I also don't have in my mind the whole picture of the folder structure. So it is hard to decide where to put what.

File: src/fe-small-advertisements/src/actions/index.ts (hunk 3 above)
Comment (AdrianValdes) [reply src/fe-small-advertisements/src/actions/index.ts:1]: true. Even the linter complains sometimes when one uses `name` in some places.

File: src/fe-small-advertisements/src/actions/types.ts (hunk 4 above)
Comment (AdrianValdes) [reply src/fe-small-advertisements/src/actions/types.ts:3]: I was trying to set the basis to in the future do something like: 

`export type UserAction = ChangeNameAction | ChangeAgeAction | ChangePasswordAction` 

I was just practicing the structure, but I realize it is not good to write code for no reason, my bad.

 To learn Typescript in React and how to use it with Redux  is confusing  sometimes :). Thanks for guidance!

File: src/fe-small-advertisements/src/actions/name.ts (hunk 5 above)
Comment (AdrianValdes) [reply src/fe-small-advertisements/src/actions/name.ts:16]: Yes, yes! I am telling you, Typescript in React still looks like a completely different language to me. I refactored the whole thing now.

File: src/fe-small-advertisements/src/actions/name.ts (hunk 7 above)
Comment (AdrianValdes) [reply src/fe-small-advertisements/src/actions/name.ts:12]: It makes sense, it look cleaner. With `redux-toolkit` one can make it look even cleaner.

File: src/fe-small-advertisements/src/components/App.tsx (hunk 9 above)
Comment (AdrianValdes) [reply src/fe-small-advertisements/src/components/App.tsx:32]: Right, we are never going to have default exports for the components, right? I have being reading about the advantages of it

File: src/fe-small-advertisements/src/index.tsx (hunk 10 above)
Comment (AdrianValdes) [reply src/fe-small-advertisements/src/index.tsx:11]: done

File: src/fe-small-advertisements/src/components/App.tsx (hunk 11 above)
Comment (AdrianValdes) [reply src/fe-small-advertisements/src/components/App.tsx:13]: I agree with you. I also think it is not necessary to create an extra function to control and input like these one. I did it because Typescript makes put a lot of extra code and it looks very weird to me. But I have to get used to it. I put it directly in the onChange method, it is not that bad.

**Comment by AdrianValdes** [comment]
Hi @paschalidi, I made all the requested changes. I have a question, where should I put a comment like the one I put here now. I wrote as an answer to something you said, but there it feels like it could be lost:

COMMENT:
For this commit I created a /features/person folder and put the personSlice.ts there. I just to put it somewhere until I read more about DDD, is that fine? Or is that sloppy of me?

Sometimes I am not sure what to do in cases like this one: On the one hand, you asked me to build just a showcase and I have to decide how to make that showcase the best way, but I also don't have in my mind the whole picture of the folder structure. So it is hard to decide where to put what.

File: src/fe-small-advertisements/src/App.tsx (hunk 2 above)
Comment (paschalidi) [reply src/fe-small-advertisements/src/App.tsx]: right. you right this wasnt clear. but we discuss the file structure we want the project to have in general. Now if I say this is alright and we merge this means also we will continue with this structure as it is now. 

So from my point of view I will have to bring this on your attention. Now you right that for this case it doesnt really matter. We can leave it like this however you will have to change it a little later. So up to you what you prefer. 

I prefer to merge on `main` only code that is production level.

File: src/fe-small-advertisements/src/actions/types.ts (hunk 4 above)
Comment (paschalidi) [reply src/fe-small-advertisements/src/actions/types.ts]: You will be fine. Keep the good work up dude. You doind fantasic

File: src/fe-small-advertisements/src/components/App.tsx (hunk 11 above)
Comment (paschalidi) [reply src/fe-small-advertisements/src/components/App.tsx]: you dont _always_ have to annotate. Is most useful to annotate modules that are reusable and have interfaces so that when we refactor something we have clean interfaces. In this case not annotating wont change that much imo.

File: src/fe-small-advertisements/src/App.tsx [hunk 12]
```diff
@@ -1,11 +1,34 @@
...
+import { useDispatch, useSelector } from 'react-redux';
+import { RootState } from './reducers';
+import { setName } from './features/person/personSlice';
+
+export const App = () => {
+  const name = useSelector((state: RootState) => state.person);
+  const dispatch = useDispatch();
+  const [newName, setNewName] = useState('');
+
+  const handleUpdateTitle = (e: FormEvent<HTMLFormElement>): void => {
```
Comment (paschalidi) [inline src/fe-small-advertisements/src/App.tsx:11]:
```suggestion
  const handleTitleUpdate = (e: FormEvent<HTMLFormElement>): void => {
```

File: src/fe-small-advertisements/src/features/person/personSlice.ts [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+  initialState: 'LeBron James',
+  reducers: {
+    setName: (state, { payload }: PayloadAction<{ name: string }>) =>
+      payload.name,
+  },
+});
+
+export const { setName } = personSlice.actions;
+
+export default personSlice.reducer;
```
Comment (paschalidi) [inline src/fe-small-advertisements/src/features/person/personSlice.ts:14]: 🚓 
```suggestion
export const personReducer = personSlice;
```
This is a good oportunity, I believe, to create an issue to configure the eslint in such way so that when we export default an eslint throws an error?

File: src/fe-small-advertisements/src/index.tsx [hunk 14]
```diff
@@ -1,12 +1,22 @@
...
+import { configureStore } from '@reduxjs/toolkit';
+import { rootReducer } from './reducers';
 
-import App from './App';
+import { App } from './App';
 import reportWebVitals from './reportWebVitals';
 
+const store = configureStore({
+  reducer: rootReducer,
+  devTools: process.env.NODE_ENV !== 'production',
```
Comment (paschalidi) [inline src/fe-small-advertisements/src/index.tsx:12]: 👍

File: src/fe-small-advertisements/src/reducers/index.ts [hunk 15]
```diff
@@ -0,0 +1,7 @@
+import personReducer from '../features/person/personSlice';
+
+export const rootReducer = {
+  person: personReducer,
+};
+
+export type RootState = typeof rootReducer;
```
Comment (paschalidi) [inline src/fe-small-advertisements/src/reducers/index.ts:7]: If you look closely to other example you might find out that in domain driven file structure usually dont have a `reducers` folder either. You might deleting it from here as well please?

**Review by paschalidi — COMMENTED** [review]
Hey good work few more comments and we good to go!

**Comment by paschalidi** [comment]
Can you also rebase this please?

File: src/fe-small-advertisements/src/features/person/personSlice.ts (hunk 13 above)
Comment (AdrianValdes) [reply src/fe-small-advertisements/src/features/person/personSlice.ts:14]: We would like this rule just for the frontEnd, or in the whole project?

Comment (AdrianValdes) [reply src/fe-small-advertisements/src/features/person/personSlice.ts:14]: I added to the whole project.

File: src/fe-small-advertisements/src/reducers/index.ts (hunk 15 above)
Comment (AdrianValdes) [reply src/fe-small-advertisements/src/reducers/index.ts:7]: I deleted it and tried to do a DDD with what we have.

**Comment by AdrianValdes** [comment]
> Can you also rebase this branch please?

Yes, I am reading about how to do that correctly. I am a bit afraid of doing something like that. But it is good to learn.

**Comment by AdrianValdes** [comment]
Hi @paschalidi , I reviewed and fixed all your commented and did a rebase. I double checked everything and I couldn't find any mistakes. I tripled checked and as far as I understand the rebase was done correctly.

Notes:
I am wondering if I should learn how to use Git Kraken or at least another tool, to do a rebase just with the command line is extremely confusing. 

All these steps are new to me, I am very grateful I am learning how to do it, but I am a bit afraid of screwing up, I don't want to make you work more than I am already doing.

**Comment by paschalidi** [comment]
> I am wondering if I should learn how to use Git Kraken or at least another tool, to do a rebase just with the command line is extremely confusing.

I told you what I do. You can decide what you want to learn and what not. 

In my experience I didnt get the impression that people thought that using a tool will make a worse developer. Although at the begining I was afraid they might do. I have worked with better developers than me in higher positions and they were using similar tooling. Its a tool that makes life easier. In the end we developers build tools for people to make their lifes easier

File: packages/eslint-config-small-ads/index.js [hunk 16]
```diff
@@ -23,6 +23,7 @@ module.exports = {
   plugins: ['react-hooks', '@typescript-eslint/eslint-plugin'],
   rules: {
     'import/prefer-default-export': 'off',
+    'import/no-default-export': 'error',
```
Comment (paschalidi) [inline packages/eslint-config-small-ads/index.js:26]: Cool!

Comment (AdrianValdes) [reply packages/eslint-config-small-ads/index.js:26]: :)

**Review by paschalidi — APPROVED** [review]
Good work Adrian!! Keep it up dude!!
